src/downloading_thread.py

```diff
@@ -0,0 +1,226 @@
+            for d in dirs:
+                os.makedirs(os.path.join(dest_path, d), exist_ok=True)
+
+            # Move each file to the destination directory
+            for file in files:
+                src_file_path = os.path.join(root, file)
+                dest_file_path = os.path.join(dest_path, file)
+                os.rename(src_file_path, dest_file_path)
+                # Update the sync map
+                keypath = self.filepath.getRemotePath(os.path.relpath(dest_file_path, dest))
+                self.updateSyncMap(keypath, metadata[keypath].timestamp)
+
+            # Cleanup: Remove the empty source directory after moving files
+            if rel_path != '.':
+                try:
+                    os.removedirs(root)
+                except OSError:
+                    # list of files is not empty
+                    logger.error(f"Files Not Empty: {os.listdir(root)}")
+                    logger.error(f"Failed to remove directory: {root}")
+
+
+    def addTask(self):
+        '''
+        add the task to the sync map
+        '''
+        self.eventQueue.put(True)
+
+    def updateSyncMap(self, path, timestamp):
+        '''
+        update the sync map with the path and timestamp
+        '''
+        self.syncMap[path] = timestamp
+
+    def downloadFile(self, path):
+        '''
+        download the file from dropbox
+        '''
+        logger.info(f"Downloading file {path}")
+        print(f"Downloading file {path}")
+        try:
+            os.makedirs(os.path.dirname(self.filepath.getTmpPath(path)), exist_ok=True)
+            self.dbx.download(path, self.filepath.getTmpPath(path))
+        except Exception as e:
+            logger.error(f"Error downloading file {path}: {e}")
+            return -1
+        return 0
+
+    def getDownLoadList(self, metadata:dict):
+        '''
+        get the list of files to download
+        '''
+        localTimes = self.retrieveLocalMTime()
+        dlist = self.retrieveDownloadList(metadata, localTimes)
+        return dlist
+
+    def retrieveLocalMTime(self):
+        '''
+        retrieve the local modified time
+        '''
+        ans = {}
+        for root, dirs, files in os.walk(self.rootDir):
+            for file in files:
+                path = os.path.join(root, file)
+                keypath = os.path.relpath(path, self.rootDir)
+                ans[self.filepath.getRemotePath(keypath)] = os.path.getmtime(path)
+        return ans
+    
+    def retrieveDownloadList(self, metadata:dict, localFiles:dict):
+        '''
+        retrieve the list of files to download
+        '''
+        dList = []
+        for k, v in metadata.items():
+            if k not in localFiles:
+                # download the file
+                dList.append(k)
+            else:
+                if datetime.timestamp(v.timestamp) > localFiles[k]:
+                    # download the file
+                    dList.append(k)
+
```

Comment:
The files_list_folder_continue interface provided by dropbox sdk seems simpler working with cursor. Could probably consider that in the last milestone as long as the current one works.